packages/@aws-cdk/cdk/lib/physical-name.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * Represents the human-readable, physical name of a resource.
+ * These can be known at code runtime,
+ * in the case of fixed, customer-provided names,
+ * at synthesis time, in case of automatically assigned names by the framework,
+ * or only at deploy time, assigned by CloudFormation -
+ * which is the default.
+ *
+ * @see #deployTime()
+ * @see #deployTimeOrAssigned()
+ * @see #fixed()
+ */
+export abstract class PhysicalName {
+  /**
+   * A physical name that will be automatically assigned at deploy time.
+   * This is the default you don't specify a physical name.
+   */
+  public static deployTime(): PhysicalName {
```

Comment:
I have to say I am struggling with these names and API, and it's really important to get right so bare with me...

We should first determine if the API should reflect "policy" or "mechanism". As in, do we want this API to tell users *when* it should be used (= policy) or *what* it does (= mechanism). In general, we prefer mechanism, which is in-line with the semantics currently implemented. 

But I think the `deployTimeOrAssigned` terminology is not clear enough, so maybe by mixing mechanism and policy we can offer something a bit more readable.

How about merging `deployTime` and `deployTimeOrAssigned` into a single `auto` method:

```ts
PhysicalName.auto(); // deploy-time (default)
PhysicalName.auto({ crossEnv: true }); // deploy-time-or-assigned
```

This will give us some future freedom, that I sense we might need. For example, we could allow configuring the default for `crossEnv` at the app level, so people can say: "I want all resources in my app to be able to be cross-env referenced by default instead of having to explicitly changing all their physical names to `deployTimeOrAssigned`".

We could also add other options to customize how name generation works:

```ts
PhysicalName.auto({ prefix: 'my-app-' })
```

What do you think?

Reply:
So, I think the key here is to make sure the customers understand what are the different options, and what consequences they have.

That's why I've chosen the names that I did (although I agree `deployTimeOrAssigned` is not the greatest name) - I think describing the options as "the name is know either statically, or at synth time (after Token resolution), or only at deploy time" is not a bad way to frame it.

I'm not sure `PhysicalName.auto({ crossEnv: true })` will make it obvious the difference between that, and the default `PhysicalName.auto()`. `PhysicalName.auto({ prefix: 'my-app-' })` is even worse in my mind, because there's no way we'll be able to do that with CFN, so this will imply physical names, which seems very non-obvious.

I do agree that having an optional props argument to the static factory method(s) is probably a nice way to future proof them.

Reply:
I do have a feeling that since there will be cases where users won't have control over how underlying resources are configured, and those will be configured without `cross: true`, then they would want to be able to override the default "auto" behavior. That's why I am proposing to merge `deployTime` and `deployTimeOrAssigned` into a single method with config. This will enable us to provide facilities for users to basically modify the default behavior of some scope:

```ts
PhysicalName.allowCrossEnvWithinThisScope(this); // this is "Construct"

new Bucket(this, 'MyBucket'); // implies `deploy-time-or-assigned`.
```

Reply:
I'm not sure that requiring customers to invoke something like `PhysicalName.allowCrossEnvWithinThisScope(this);` is a great experience...

I think you raise an interesting point though: perhaps the default physical name should be `deployTimeOrAssigned()`, or, with your proposal, `auto({ crossEnv: true })`(in other words, should `crossEnv` be `true` by default in the call to `auto`)?

Reply:
FWIW, Aspects could totally cover the use case you're thinking of there, Elad.

Reply:
Changed to `PhysicalName.auto({ generateIfCrossEnv: true })`. Let me know if this works.